wts/tokens.py

```diff
@@ -1,37 +1,90 @@
 import flask
-import requests
 import time
+import httpx
 
-from cdiserrors import AuthError, InternalError
+from cdiserrors import AuthError, InternalError, UserError
 
 from .models import db, RefreshToken
 from .utils import get_oauth_client
 
 
+def get_data_for_fence_request(refresh_token):
+    """
+    Given `refresh_token`, prepare data for request to IDP's token endpoint.
+
+    Args:
+        refresh_token (wts.models.RefreshToken): refresh token reference
+    """
+    token = refresh_token.token
+    token_bytes = bytes(token, encoding="utf-8")
+    token = flask.current_app.encryption_key.decrypt(token_bytes).decode("utf-8")
+    client = get_oauth_client(idp=refresh_token.idp)
+    url = client.metadata.get("access_token_url")
+    data = {"grant_type": "refresh_token", "refresh_token": token}
+    auth = (client.client_id, client.client_secret)
+    return url, data, auth
+
+
 def get_access_token(requested_idp, expires=None):
-    client = get_oauth_client(idp=requested_idp)
-    now = int(time.time())
+    if requested_idp not in flask.current_app.oauth2_clients:
+        raise UserError('Requested IDP "{}" is not configured'.format(requested_idp))
```

Comment:
this is nitpicky, but I think the standard capitalization of identity provider is `IdP`